libgalois/include/galois/ParallelSTL.h

```diff
@@ -305,6 +305,60 @@ std::enable_if_t<!std::is_scalar<internal::Val_ty<I>>::value> destroy(I first,
 template <class I>
 std::enable_if_t<std::is_scalar<internal::Val_ty<I>>::value> destroy(I, I) {}
 
+/**
+ * Does a partial sum from first -> last and writes the results to the d_first
+ * iterator.
+ */
+template <class InputIt, class OutputIt>
+OutputIt partial_sum(InputIt first, InputIt last, OutputIt d_first) {
+  using ValueType = typename std::iterator_traits<InputIt>::value_type;
+
+  size_t sizeOfVector = std::distance(first, last);
+
+  const size_t blockSize = 1 << 20;
+  // optimization possible here: if num blocks = 1, then no need for 2 passes
+  const size_t numBlocks = (sizeOfVector + blockSize - 1) / blockSize;
```

Comment:
I think the number of blocks should be the same as the number of threads. There's no need to load balance here and the iterations of the serial loop would be reduced.

Reply:
Doable.

My concern is that for trivially small vectors there really isn't a point going multi-threaded since there is overhead with parallelism involved, but I think in most cases it probably big enough of an issue to design around that.

One more comment I forgot to make: the std::partial_sum interface makes it so you write to a new vector (the output vector), so doing this actually adds another O(n) memory overhead. It probably isn't too much of a deal considering relabeling requires having 2 graphs in memory anyways.

Reply:
Made the change; do check to make sure logic is fine.